EssentialFeed/EssentialFeedTests/Comments/API/LoadCommentsFromRemoteUseCasesTests.swift

```diff
@@ -0,0 +1,152 @@
+	func test_load_deliversErrorOn200HTTPResponeWithInvalidJSON() {
+		let (sut, client) =  makeSUT()
+		expect(sut, toCompleteWith: failure(.invalidData), when: {
+			let invalidJSON = Data("invalid json".utf8)
+			client.complete(withStatusCode: 200, data: invalidJSON)
+		})
+	}
+	
+	func test_load_deliversNoItemsOn200HTTPReponseWithEmptyJSONList() {
+		let (sut, client) = makeSUT()
+		expect(sut, toCompleteWith: .success([]), when: {
+			let emptyListJSON = makeItemsJSON([])
+			client.complete(withStatusCode: 200, data: emptyListJSON)
+		})
+	}
+	
+	func test_load_deliversItemsOn200HTTPResponseWithJSONItems() {
+		let (sut, client) = makeSUT()
+		
+		let author = Author(username: "a username")
+		let item1 = makeItem(id: UUID(), author: author)
+		
+		let item2 = makeItem(id: UUID(), author: author)
+		
+		let items = [item1.model, item2.model]
+		
+		expect(sut, toCompleteWith: .success(items), when: {
+			let json = makeItemsJSON([item1.json, item2.json])
+			client.complete(withStatusCode: 200, data: json)
+		})
+	}
+	
+	func test_load_doesNotDeliverResultAfterSUTInstanceHasBeenDeallocated() {
+		let client = HTTPClientSpy()
+		let url = URL(string: "http://any-url.com")!
+		var sut: RemoteCommentLoader? = RemoteCommentLoader(client: client, url: url)
+		
+		var capturedResults = [CommentLoader.Result]()
+		sut?.load { capturedResults.append($0) }
+		
+		sut = nil
+		client.complete(withStatusCode: 200, data: makeItemsJSON([]))
+		
+		XCTAssertTrue(capturedResults.isEmpty)
+	}
+	
+	// MARK: - Helpers
+	
+	private func makeSUT(url: URL = anyURL(), file: StaticString = #filePath, line: UInt = #line) -> (sut: RemoteCommentLoader, client: HTTPClientSpy) {
+		let client =  HTTPClientSpy()
+		let sut = RemoteCommentLoader(client: client, url: url)
+		trackForMemoryLeaks(client)
+		trackForMemoryLeaks(sut)
+		
+		return (sut, client)
+	}
+	
+	private func failure(_ error: RemoteCommentLoader.Error) -> CommentLoader.Result {
+		return .failure(error)
+	}
+
+	private func expect(_ sut: RemoteCommentLoader, toCompleteWith expectedResult: CommentLoader.Result, when action: () -> Void, file: StaticString = #filePath, line: UInt = #line) {
+		
+		let exp = expectation(description: "Wait for load completion")
+		sut.load { receivedResult in
+			switch (receivedResult, expectedResult) {
+			case let (.success(receivedItems), .success(expectedItems)):
+				XCTAssertEqual(receivedItems, expectedItems, file: file, line: line)
+			case let (.failure(receivedError as RemoteCommentLoader.Error), .failure(expectedError as RemoteCommentLoader.Error)):
+				XCTAssertEqual(receivedError, expectedError, file: file, line: line)
+			default:
+				XCTFail("Expected result \(expectedResult) got \(receivedResult) instead", file: file, line: line)
+			}
+			exp.fulfill()
+		}
+		
+		action()
+		wait(for: [exp], timeout: 1.0)
+	}
+	
+	private func makeItem(id: UUID, message: String? = "", createdAt: Date? = Date(), author: Author) -> (model: Comment, json: [String: Any]) {
```

Comment:
The `message` and the `createdAt` params shouldn't be optional. If someone passes `nil`, it'd crash.

```suggestion
	private func makeItem(id: UUID, message: String = "", createdAt: Date = Date(), author: Author) -> (model: Comment, json: [String: Any]) {
```

Reply:
done